diff --git a/mesonbuild/compilers/compilers.py b/mesonbuild/compilers/compilers.py
index 424bcc19b..a1f4929f1 100644
--- a/mesonbuild/compilers/compilers.py
+++ b/mesonbuild/compilers/compilers.py
@@ -47,7 +47,7 @@ Also add corresponding autodetection code in detect.py."""
 header_suffixes = {'h', 'hh', 'hpp', 'hxx', 'H', 'ipp', 'moc', 'vapi', 'di'}
 obj_suffixes = {'o', 'obj', 'res'}
 # To the emscripten compiler, .js files are libraries
-lib_suffixes = {'a', 'lib', 'dll', 'dll.a', 'dylib', 'so', 'js'}
+lib_suffixes = {'a', 'lib', 'dll', 'dll.a', 'dylib', 'so', 'x', 'js'}
 # Mapping of language to suffixes of files that should always be in that language
 # This means we can't include .h headers here since they could be C, C++, ObjC, etc.
 # First suffix is the language's default.
