diff --git a/mesonbuild/linkers/linkers.py b/mesonbuild/linkers/linkers.py
index 176fb3348..f0fd70ce0 100644
--- a/mesonbuild/linkers/linkers.py
+++ b/mesonbuild/linkers/linkers.py
@@ -394,6 +394,10 @@ class ArLinker(ArLikeLinker, StaticLinker):
             return self.std_args
 
 
+class ZOSArLinker(ArLinker):
+    id = "zar"
+
+
 class AppleArLinker(ArLinker):
 
     # mostly this is used to determine that we need to call ranlib
@@ -1571,6 +1575,13 @@ class AIXDynamicLinker(PosixDynamicLinkerMixin, DynamicLinker):
         return ['-pthread']
 
 
+class ZOSDynamicLinker(PosixDynamicLinkerMixin, DynamicLinker):
+
+    """z/OS ld implementation."""
+
+    id = "ld.zos"
+
+
 class OptlinkDynamicLinker(VisualStudioLikeLinkerMixin, DynamicLinker):
 
     """Digital Mars dynamic linker for windows."""
