diff --git a/mesonbuild/build.py b/mesonbuild/build.py
index 3c9d3a06c..2263e7079 100644
--- a/mesonbuild/build.py
+++ b/mesonbuild/build.py
@@ -2317,6 +2317,15 @@ class SharedLibrary(BuildTarget):
             suffix = 'so'
             # Android doesn't support shared_library versioning
             self.filename_tpl = '{0.prefix}{0.name}.{0.suffix}'
+        elif self.environment.machines[self.for_machine].is_zos():
+            if len(self.name) <= 8 and self.name.isupper() and not self.prefix and not self.suffix:
+                prefix = ''
+                self.filename_tpl = '{0.name}'
+            else:
+                prefix = 'lib'
+                suffix = 'so'
+                self.filename_tpl = '{0.prefix}{0.name}.{0.suffix}'
+            import_filename_tpl = '{0.prefix}{0.name}.x'
         else:
             prefix = 'lib'
             suffix = 'so'
@@ -2685,7 +2694,7 @@ class CustomTarget(Target, CustomTargetBase, CommandBase):
             return self.depfile
 
     def is_linkable_output(self, output: str) -> bool:
-        if output.endswith(('.a', '.dll', '.lib', '.so', '.dylib')):
+        if output.endswith(('.a', '.dll', '.lib', '.so', '.dylib', '.x')):
             return True
         # libfoo.so.X soname
         if re.search(r'\.so(\.\d+)*$', output):
