diff --git a/mesonbuild/envconfig.py b/mesonbuild/envconfig.py
index 86bad9be2..a1397bc33 100644
--- a/mesonbuild/envconfig.py
+++ b/mesonbuild/envconfig.py
@@ -64,6 +64,19 @@ known_cpu_families = (
     'wasm64',
     'x86',
     'x86_64',
+    # z/Architecture CPU Families
+    '3932',
+    '3931',
+    '8562',
+    '8561',
+    '3907',
+    '3906',
+    '2965',
+    '2964',
+    '2828',
+    '2827',
+    '2818',
+    '2817',
 )
 
 # It would feel more natural to call this "64_BIT_CPU_FAMILIES", but
@@ -81,6 +94,19 @@ CPU_FAMILIES_64_BIT = [
     'sw_64',
     'wasm64',
     'x86_64',
+    # z/Architecture CPU Families
+    '3932',
+    '3931',
+    '8562',
+    '8561',
+    '3907',
+    '3906',
+    '2965',
+    '2964',
+    '2828',
+    '2827',
+    '2818',
+    '2817',
 ]
 
 # Map from language identifiers to environment variables.
@@ -362,6 +388,12 @@ class MachineInfo(HoldableObject):
         """
         return self.system == 'aix'
 
+    def is_zos(self) -> bool:
+        """
+        Machine is z/OS?
+        """
+        return self.system == 'os/390'
+
     def is_irix(self) -> bool:
         """Machine is IRIX?"""
         return self.system.startswith('irix')
